### ISSUE TITLE
IRC notifications template is the `show` output of a string

### ISSUE THREAD

**Issue by RyanGlScott** [haskell-CI/haskell-ci#269]
I noticed a [suspicious change](https://github.com/ekmett/semigroups/commit/6c3dcbed31b3ba1ef3ee66126a40758d1d8f452a#diff-354f30a63fb0907d4ad57269548329e3L21) in a recent update to `semigroups`' `.travis.yml` file:

```diff
-      - "\x0313semigroups\x03/\x0306%{branch}\x03 \x0314%{commit}\x03 %{build_url} %{message}"
-
+      - "\"\\x0313semigroups\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}\""
```

It appears that the previous string was fed into `show`, resulting in everything being quoted more than it should. I didn't check to see if this notification actually showed up on the #haskell-lens IRC channel, but my guess it that it wouldn't have.

**Comment by RyanGlScott** [comment]
[Another part of the `.travis.yml` file](https://github.com/ekmett/semigroups/commit/6c3dcbed31b3ba1ef3ee66126a40758d1d8f452a#diff-354f30a63fb0907d4ad57269548329e3L18) actually went in the opposite direction in terms of how much quoting is present:

```diff
-      - "irc.freenode.org#haskell-lens"
+      - irc.freenode.org#haskell-lens
```

**Comment by RyanGlScott** [comment]
I tried fixing this recently, but there appears to be more going on here than I originally thought. I discovered the line of code where the `notifications` template lives:

https://github.com/haskell-CI/haskell-ci/blob/2bf66a49942b1eb79c31cc71743df51d5950ebff/src/HaskellCI/Travis.hs#L382

As an first attempt to start fixing this, I made a small change to remove that first explicit double quote:

```diff
-                    [ "\"\\x0313" ++ projectName ++ "\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}\""
+                    [ "\\x0313" ++ projectName ++ "\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}\""
```

Seems simple enough, right? But here's how this change manifests in the generated `.travis.yml` file:

```diff
-      - "\"\\x0313semigroups\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}\""
+      - \x0313semigroups\x03/\x0306%{branch}\x03 \x0314%{commit}\x03 %{build_url} %{message}"
```

Huh? Even though I only removed the characters `\"`, that caused _both_ double quotes to disappear from the start of the template, as well as the `\` character that followed the `\"`!

A similar phenomenon occurs if I try to explicitly escape the `channels` names. If I make this change:

```diff
-                { tiChannels = cfgIrcChannels
+                { tiChannels = map show cfgIrcChannels
```

Then that change manifests in the generated `.travis.yml` file like so:

```diff
-      - irc.freenode.org#haskell-lens
+      - "\"irc.freenode.org#haskell-lens\""
```

Again, this is quite baffling. Somehow, there are now _two_ sets of double quotes surrounding the channel name!

This leads me to believe that there is some other part of `haskell-ci` that is adding an extra level of escaping to this YAML output, but I'm not familiar enough with the codebase to determine where that could be.

**Comment by mithrandi** [comment]
The YAML snippet of:
```- irc.freenode.org#haskell-lens```
is equivalent to
```- "irc.freenode.org#haskell-lens"```

YAML string scalars can be written without the quotes as long as they are limited to certain characters; but they are necessary if you want to use \-escapes, leading/trailing whitespace, etc.

**Comment by RyanGlScott** [comment]
The culprit appears to be the `literal` function in `HaskellCI.YamlSyntax`:

https://github.com/haskell-CI/haskell-ci/blob/0ebd7030064aa4c3500ebc85a3d20fc44cf6fe04/src/HaskellCI/YamlSyntax.hs#L264-L274

This determines whether a string needs to be escaped with double quotes or not. `literal` believes that `"\"\\x0313semigroups\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}\""` requires escaping, but `"\\x0313semigroups\\x03/\\x0306%{branch}\\x03 \\x0314%{commit}\\x03 %{build_url} %{message}"` does not. Perhaps this is a sign that using this in YAML would actually just work?

```yaml
    template:
      - \x0313semigroups\x03/\x0306%{branch}\x03 \x0314%{commit}\x03 %{build_url} %{message}
```

I haven't tried this, but if it _does_ work, then my attempted fix in https://github.com/haskell-CI/haskell-ci/issues/269#issuecomment-488287573 might work after all.

**Comment by RyanGlScott** [comment]
I experimented with the idea in https://github.com/haskell-CI/haskell-ci/issues/269#issuecomment-586318569 in ekmett/semigroups@6d12678f0be4ae84b41d7efd78d6b706a6a19c6c. At the very least, Travis was able to parse the new template, since https://travis-ci.org/ekmett/semigroups/builds/650471628 succeeded. On the other hand, I wasn't able to tell if the `travis-ci` IRC bot renders the new template correctly, since `travis-ci` doesn't appear to respond to commits to non-`master` branches on the `#haskell-lens` IRC channel.

@phadej, do you think we should just use the new template style going forward. I don't see any risk in doing so, especially since the current template is already broken :)

**Comment by phadej** [comment]
Could you try https://github.com/haskell-CI/haskell-ci/tree/try-to-fix-irc-template branch?

**Comment by RyanGlScott** [comment]
I'm not sure what you mean. The change you made in da30c5c516d076fb91b4e401487c5f64d8b9f54c was exactly the same change I made when experimenting in ekmett/semigroups@6d12678 (modulo bumping the `haskell-ci` version).

**Comment by phadej** [comment]
@RyanGlScott I understood it should work, shouldn't it?

**Comment by RyanGlScott** [comment]
I believe it should work, yes. The only problem is that I don't have a reliable way to test it "in the wild" (i.e., observe the `travis-ci` IRC bot post a message with the new template), since AFAICT that would require pushing these changes to the `master` branch of some project. Perhaps I should just do that, but it might be wise to associate this `haskell-ci` change with a new development version number first so that I have something to blame in case it goes wrong :)

**Comment by phadej** [comment]
that commit has new development version :)

**Comment by RyanGlScott** [comment]
Alright, I decided to be bold and just push a change to the `master` branch of `semigroups` in ekmett/semigroups@c4524786f02debd2422355b4b0aecd05dce2b84f. Let's see if the `travis-ci` bot responds to that in `#haskell-lens`...

**Comment by phadej** [comment]
@RyanGlScott thanks, I'm holding my breath now :)

**Comment by phadej** [comment]
Fixed in #362
